projects/common-platform-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/EventService.kt

```diff
@@ -0,0 +1,234 @@
+    ): InsertEventResult =
+        audit(BusinessInteractionCode.INSERT_EVENT) { audit ->
+
+            // 1. Create and save the event entity
+            val savedEvent = eventRepository.save(
+                Event(
+                    id = null,
+                    person = person,
+                    number = eventRepository.getNextEventNumber(person.id!!),
+                    referralDate = sittingDay.toLocalDate(), // TODO: Identify event's referral date
+                    active = true,
+                    ftcCount = 0,
+                    notes = "caseUrn:$caseUrn"
+                )
+            )
+
+            val hoOffenceCode =
+                hearingOffence.offenceCode?.let {
+                    detailedOffenceRepository.findByCode(it)?.homeOfficeCode?.replace(
+                        "/",
+                        ""
+                    )
+                }
+                    ?: throw IllegalArgumentException("Home Office Code cannot be null")
+
+            // 2. Create the main offence record from the hearing message
+            val savedMainOffence = mainOffenceRepository.save(
+                MainOffence(
+                    id = null,
+                    date = LocalDate.now(), // TODO: From the offence -> Wording free text field
+                    event = savedEvent,
+                    offence = offenceRepository.findOffence(hoOffenceCode),
+                    softDeleted = false,
+                    count = 1, // TODO: Need to identify how offence count is used, in most cases appears to be 1
+                    person = person,
+                )
+            )
+
+            val court = courtRepository.getByOuCode(courtCode)
+            val courtAppearanceContactType = contactTypeRepository.getByCode(ContactTypeCode.COURT_APPEARANCE.code)
+            val unallocatedTeam = teamRepository.findByCode(court.provider.code + "UAT")
+            val unallocatedStaff = staffRepository.findByCode(unallocatedTeam.code + "U")
+            val trialAdjournmentRefData = referenceDataRepository.trialAdjournmentAppearanceType()
+            val remandedInCustodyOutcome = referenceDataRepository.remandedInCustodyOutcome()
+            val remandedInCustodyStatus = referenceDataRepository.remandedInCustodyStatus()
+            val plea = when (hearingOffence.plea?.pleaValue) {
+                "NOT_GUILTY" -> referenceDataRepository.findByCodeAndDatasetCode(
+                    ReferenceData.StandardRefDataCode.NOT_GUILTY.code,
+                    DatasetCode.PLEA
+                )
+
+                "GUILTY" -> referenceDataRepository.findByCodeAndDatasetCode(
+                    ReferenceData.StandardRefDataCode.GUILTY.code,
+                    DatasetCode.PLEA
+                )
+
+                else -> referenceDataRepository.findByCodeAndDatasetCode(
+                    ReferenceData.StandardRefDataCode.NOT_KNOWN_PLEA.code,
+                    DatasetCode.PLEA
+                )
+            }
```

Comment:
It might be possible to simplify this code a bit, for example:

```suggestion
            val plea = referenceDataRepository.findByCodeAndDatasetCode(
                when (hearingOffence.plea?.pleaValue) {
                    "NOT_GUILTY" -> ReferenceData.StandardRefDataCode.NOT_GUILTY.code
                    "GUILTY" -> ReferenceData.StandardRefDataCode.GUILTY.code
                    else -> ReferenceData.StandardRefDataCode.NOT_KNOWN_PLEA.code
                }, 
                DatasetCode.PLEA)
```